Limit script

CorentinBT opened the pull request:
Added a script to apply network and/or memory restrictions to a process and children of this process. This is useful to watch the behavior of a workload under some limitations.

It uses the traffic control, cgroup and iptables utilities of Linux. The libcgroup-tools library is needed as well to correctly manage cgroups . 

To restrict the physical amount of memory, the script simply creates a new cgroup folder (usually in `/sys/fs/cgroup/memory`) and changes the value in`memory.limit_in_bytes` file to the desired limit value. The process is then associated to this cgroup by appending his pid to the `tasks` file. 

To track the network usage of a particular process, the script creates an iptables rule to mark all packets sent and received by the process. 

For the latency and upload limit, the script creates two new queue discipline. The `htb` discipline which allows to redirect marked packets, using a filter, to the  `netem` discipline which adds delay to outgoing packets. An intermediate `htb` class, between these two queue disciplines, is also created to limit the incoming  packet to a desired rate (the upload rate). For example with a delay of `50ms` and upload limit of `1mbit` , the queue tree can be visualized as : 
<p align="center"> 
<img src="https://user-images.githubusercontent.com/32176761/44657386-f3af3a00-a9fc-11e8-83f0-a0bea57e3bb9.png">
</p>

The traffic control utility cannot be used to control ingress traffic. Thus for the download limit, the script uses an iptables module called `hashlimit` which drops incoming packets that exceed a desired limit (the download rate).

amete commented:
Hi @CorentinBT,

Thanks a lot for this PR and apologies for taking too long. I have some first basic comments:

- I think we can update `prmon/package/scripts` to install this script during `make install`
- I think it'd be nicer to use newlines a bit less liberally in the help message
- I tried doing `./cg_limit.sh -b amete -m 1G -- mem-burner -m 2000` only to get `runuser: invalid option -- 'u'`. I'm using SLC6 w/ `runuser (GNU coreutils) 8.4` which doesn't seem to have a `-u` option for runuser looking through its help. I guess we need to make the script `runuser` version aware if we want to support older builds.
- Working around the error above, I could move forward only to have an error that says `/afs/cern.ch/user/a/amete/.bash_profile: Permission denied`. This makes sense, as my home is under `afs` but I'm running as `root` (on a specific machine) that is not allowed to read/write there. We need to keep this in mind. Maybe I'm not using the script quite right?

Best,
Serhan

CorentinBT commented:
Hello @amete. 

Thank you for your review. 

- Yes but should we check with CMake if the user has the proper libraries to run this script?
- You are right, I have removed all the useless newlines.
- I have replaced the runuser command by sudo.
- The script is running as root, however the privileges are dropped to the username, provided by -b option, before running the process.  I think the problem is because I was not setting the HOME environment to the username's environment (now fixed by the -H option with  sudo cmd).

amete commented:
Hi @CorentinBT,

Thank you very much for the changes and apologies for taking ages to come back to this:

- I think we can simply install the script so that it shows up in the proper `bin` directory. It doesn't do anything other than that so not having the proper libraries won't be a real problem
- Now I'm having this problem. I compile `prmon` using the Athena environment (to pickup a more recent gcc than what SLC6 natively gives me). The binary, in this example `mem-burner` runs fine as is. However, when we do `sudo mem-burner` which is what the script internally does to fork and launch the application (to the first order), I get `GLIBCXX` version errors. I think this is because by default `sudo` doesn't conserve the environment variables/setup. I tried to fiddle around w/ the `-E` option w/o much success.  Any ideas?

Best,
Serhan

CorentinBT commented:
Hello @amete ,

Thank you for your answers.

- I added an entry to CMake to install `cg_limit.sh` script in the proper `bin` folder.

- The problem seems to affect only certain distributions, since I don't have the correct distribution to reproduce this problem maybe you should try these options with sudo : `sudo -u $UNAME -E env "PATH=$PATH" $@`

CorentinBT commented on package/scripts/cg_limit.sh:
```diff
@@ -0,0 +1,256 @@
+
+# now executing the program
+shift "$((OPTIND - 1))"
+sudo -H -u $UNAME $@
```
Nice then if the problem is resolved. I don't see any reasons that would cause conflicts. The best way to prove this would be to test this script on multiple distributions.

graeme-a-stewart commented:
Hi @CorentinBT @amete - were the discussion on this completed? We should either merge this in as an add on utility or forget it...

CorentinBT commented:
Hi @graeme-a-stewart, sorry for the delay I haven't seen your message. I would say the discussion is complete but it really depends on you, if you feel this script is related to prmon or not.

graeme-a-stewart commented:
As we still have unresolved issues with this script, we close this PR, but the proposed script is "stashed" on the `cgroups-script` branch, should there be a need to revive work on it. (https://github.com/HSF/prmon/tree/cgroups-script was created by merging the current head of `master` with https://github.com/CorentinBT/prmon.)

(Thanks to @CorentinBT for having worked on this - we will pick it up again if there is a need for this functionality.)
